File: artifacts/api-server/src/bot/messages.ts
import type { TokenConfig } from "./session";
import type { DexServicePrices } from "./dex-pricing";
import { fmtSolUsd } from "./dex-pricing";

/** Minimum SOL that must be in the wallet for the bot to operate. */
export const DEPLOYMENT_FEE = 5; // SOL

export function mainMenuMessage(): string {
  return (
    `*TokenLaunchBot* — Solana SPL Token Deployer\n\n` +
    `Deploy tokens on Solana in minutes.\n\n` +
    `*Commands*\n` +
    `/create — Start token creation\n` +
    `/wallet — Deployment wallet info\n` +
    `/withdraw — Withdraw SOL\n` +
    `/panel — Token control panel\n` +
    `/review — Review deployment details\n` +
    `/launch — Deploy your token\n` +
    `/reset — Start over\n` +
    `/help — Show help`
  );
}

export function walletMessage(
  address: string,
  balance: number,
  keyConfigured: boolean
): string {
  return (
    `*Deployment Wallet*\n\n` +
    `*Address*\n\`${address}\`\n\n` +
    `*Balance*\n\`${balance.toFixed(4)} SOL\`\n\n` +
    `*Minimum Required*\n\`${DEPLOYMENT_FEE} SOL\`\n\n` +
    `*Private Key*\n${keyConfigured ? "`Configured ✓`" : "`Not configured ✗`"}\n\n` +
    `*Network*\nSolana Mainnet\n\n` +
    `_This is a fixed deployment wallet. It cannot be changed._`
  );
}

export function panelMessage(
  mintAddress: string,
  symbol: string,
  uiBalance: number,
  decimals: number
): string {
  return (
    `*Token Control Panel*\n\n` +
    `*Token*\n\`${symbol}\`\n\n` +
    `*Mint*\n\`${mintAddress}\`\n\n` +
    `*Wallet Balance*\n\`${uiBalance.toLocaleString(undefined, { maximumFractionDigits: decimals })} ${symbol}\`\n\n` +
    `Choose an action:`
  );
}

export function burnConfirmMessage(
  symbol: string,
  uiAmount: number,
  portion: "half" | "all"
): string {
  return (
    `*Confirm Burn*\n\n` +
    `You are about to burn *${portion === "half" ? "half" : "all"}* of your ${symbol} tokens.\n\n` +
    `*Amount:* \`${uiAmount.toLocaleString()} ${symbol}\`\n\n` +
    `This is *irreversible.* The tokens will be permanently destroyed.`
  );
}

export function burnSuccessMessage(
  symbol: string,
  uiAmount: number,
  txSignature: string
): string {
  const shortTx = `${txSignature.slice(0, 8)}...${txSignature.slice(-8)}`;
  return (
    `*Tokens Burned*\n\n` +
    `*Amount:* \`${uiAmount.toLocaleString()} ${symbol}\`\n\n` +
    `*Transaction:* \`${shortTx}\`\n\n` +
    `*Status:* Confirmed`
  );
}

export function transferTokenReviewMessage(
  toAddress: string,
  uiAmount: number,
  symbol: string
): string {
  return (
    `*Transfer Review*\n\n` +
    `*To*\n\`${toAddress}\`\n\n` +
    `*Amount*\n\`${uiAmount.toLocaleString()} ${symbol}\`\n\n` +
    `Confirm to send.`
  );
}

export function transferTokenSuccessMessage(
  toAddress: string,
  uiAmount: number,
  symbol: string,
  txSignature: string
): string {
  const shortTx = `${txSignature.slice(0, 8)}...${txSignature.slice(-8)}`;
  return (
    `*Transfer Sent*\n\n` +
    `*To:* \`${toAddress}\`\n` +
    `*Amount:* \`${uiAmount.toLocaleString()} ${symbol}\`\n` +
    `*Transaction:* \`${shortTx}\`\n\n` +
    `*Status:* Confirmed`
  );
}

export function revokeConfirmMessage(
  type: "mint" | "freeze",
  symbol: string
): string {
  const label = type === "mint" ? "Mint Authority" : "Freeze Authority";
  const effect =
    type === "mint"
      ? "No more tokens can ever be minted."
      : "Token accounts can no longer be frozen.";
  return (
    `*Confirm Revoke ${label}*\n\n` +
    `Token: \`${symbol}\`\n\n` +
    `${effect}\n\n` +
    `This is *irreversible.*`
  );
}

export function revokeSuccessMessage(
  type: "mint" | "freeze",
  symbol: string,
  txSignature: string
): string {
  const label = type === "mint" ? "Mint Authority" : "Freeze Authority";
  const shortTx = `${txSignature.slice(0, 8)}...${txSignature.slice(-8)}`;
  return (
    `*${label} Revoked*\n\n` +
    `Token: \`${symbol}\`\n` +
    `*Transaction:* \`${shortTx}\`\n\n` +
    `*Status:* Confirmed`
  );
}

export function withdrawReviewMessage(
  toAddress: string,
  amount: number,
  balance: number
): string {
  return (
    `*Withdrawal Review*\n\n` +
    `*To*\n\`${toAddress}\`\n\n` +
    `*Amount*\n\`${amount} SOL\`\n\n` +
    `*Current Balance*\n\`${balance.toFixed(4)} SOL\`\n\n` +
    `*Remaining After*\n\`${(balance - amount).toFixed(4)} SOL\`\n\n` +
    `*Network*\nSolana Mainnet\n\n` +
    `Confirm to send.`
  );
}

export function withdrawSuccessMessage(
  toAddress: string,
  amount: number,
  txSignature: string
): string {
  const shortTx = `${txSignature.slice(0, 8)}...${txSignature.slice(-8)}`;
  return (
    `*Withdrawal Sent*\n\n` +
    `*To*\n\`${toAddress}\`\n\n` +
    `*Amount*\n\`${amount} SOL\`\n\n` +
    `*Transaction*\n\`${shortTx}\`\n\n` +
    `*Status:* Confirmed`
  );
}

export function helpMessage(): string {
  return (
    `*TokenLaunchBot Help*\n\n` +
    `*Supported Network:* Solana (SPL Tokens only)\n\n` +
    `*Minimum Wallet Balance:* ${DEPLOYMENT_FEE} SOL\n` +
    `_This covers token creation, creator buy, and all on-chain fees._\n\n` +
    `*How it works:*\n` +
    `1. /create — Enter token details\n` +
    `2. /review — Check everything looks right\n` +
    `3. /launch — Choose launchpad, set creator buy & market cap target\n` +
    `4. /panel — Manage your token after launch\n\n` +
    `*Required fields:* Name, Symbol, Supply, Decimals\n\n` +
    `*Optional:* Description, Logo, Website, Telegram, Twitter, authority settings\n\n` +
    `The bot uses a single dedicated deployment wallet. Use /wallet to check your balance.`
  );
}

export function reviewMessage(
  token: TokenConfig,
  walletAddress: string,
  fee: number
): string {
  const lines: string[] = [];
  lines.push(`*Deployment Review*\n`);
  lines.push(`*Token Details*`);
  lines.push(`Name: \`${token.name}\``);
  lines.push(`Symbol: \`${token.symbol}\``);
  lines.push(`Total Supply: \`${(token.supply ?? 0).toLocaleString()}\``);
  lines.push(`Decimals: \`${token.decimals ?? 9}\``);

  if (token.description) lines.push(`Description: ${token.description}`);
  if (token.logoUrl) lines.push(`Logo: ${token.logoUrl}`);
  if (token.website) lines.push(`Website: ${token.website}`);
  if (token.telegram) lines.push(`Telegram: ${token.telegram}`);
  if (token.twitter) lines.push(`Twitter: ${token.twitter}`);

  lines.push(`\n*Authority Settings*`);
  lines.push(`Revoke Mint: ${token.revokeMint ? "Yes" : "No"}`);
  lines.push(`Revoke Freeze: ${token.revokeFreeze ? "Yes" : "No"}`);

  lines.push(`\n*Deployment*`);
  lines.push(`Wallet: \`${walletAddress}\``);
  lines.push(`Min wallet balance: \`${fee} SOL\``);
  lines.push(`Network: Solana Mainnet`);

  lines.push(`\nReady to launch? Use /launch to continue.`);
  return lines.join("\n");
}

/** Simple 2-option launchpad selector — no descriptions. */
export function launchpadSelectMessage(tokenName: string, tokenSymbol: string): string {
  return (
    `*Select Launchpad*\n\n` +
    `Token: *${tokenName}* (${tokenSymbol})\n\n` +
    `Choose where to launch your coin:`
  );
}

/** Step 2: how much SOL should the creator buy on launch? */
export function creatorBuyMessage(launchpad: string, balance: number): string {
  return (
    `*Creator Buy*\n\n` +
    `Launchpad: *${launchpad === "pumpfun" ? "Pump.fun" : "Raydium"}*\n` +
    `Wallet balance: \`${balance.toFixed(4)} SOL\`\n\n` +
    `How much SOL do you want to invest as the creator?\n` +
    `_This amount will be used to buy your own token on launch._`
  );
}

/** Step 3: what market cap (USD) should trigger the auto-sell? */
export function targetMcapMessage(creatorBuySol: number): string {
  const buyStr = creatorBuySol > 0 ? `\`${creatorBuySol} SOL\`` : "_none_";
  return (
    `*Auto-Sell Target*\n\n` +
    `Creator buy: ${buyStr}\n\n` +
    `At what market cap (USD) should the bot automatically sell all your creator tokens?\n\n` +
    `_The bot monitors the market cap every 30 seconds and sells everything when the target is hit._`
  );
}

/** Step 4: DEX Screener options with live prices. */
export function dexOptionsMessage(
  prices: DexServicePrices,
  balance: number,
  creatorBuySol: number,
  targetMcap: number
): string {
  const mcapStr = targetMcap > 0
    ? `$${targetMcap.toLocaleString()}`
    : "No auto-sell";

  const lines: string[] = [
    `*DEX Screener Options* _(optional)_\n`,
    `Creator buy: \`${creatorBuySol > 0 ? creatorBuySol + " SOL" : "none"}\``,
    `Auto-sell at: \`${mcapStr}\``,
    `Wallet: \`${balance.toFixed(4)} SOL\`\n`,
    `*DEX Update* — ${fmtSolUsd(prices.updateSol, prices.updateUsd)}`,
    `_Adds/updates your token info on DEX Screener._\n`,
    `*DEX Boost* — ${fmtSolUsd(prices.boostSol, prices.boostUsd)}`,
    `_Boosts your token to the trending section._\n`,
    `Toggle the options below, then tap *Confirm & Launch*.\n`,
    `_SOL rate: $${prices.solUsd.toFixed(2)}/SOL_`,
  ];
  return lines.join("\n");
}

export function successMessage(
  mintAddress: string,
  txSignature: string,
  solscanUrl: string,
  timestamp: string
): string {
  const shortTx = `${txSignature.slice(0, 8)}...${txSignature.slice(-8)}`;
  return (
    `*Token Deployed Successfully* 🚀\n\n` +
    `*Mint Address*\n\`${mintAddress}\`\n\n` +
    `*Transaction*\n\`${shortTx}\`\n\n` +
    `*Solscan*\n${solscanUrl}\n\n` +
    `*Status:* Confirmed\n` +
    `*Time:* ${new Date(timestamp).toUTCString()}\n\n` +
    `Tap *Token Panel* to manage your token.`
  );
}

export function pumpfunSuccessMessage(
  mintAddress: string,
  txSignature: string,
  viewUrl: string,
  timestamp: string,
  creatorBuySol: number,
  targetMcapUsd: number
): string {
  const shortTx = `${txSignature.slice(0, 8)}...${txSignature.slice(-8)}`;
  const lines: string[] = [
    `*Token Live on Pump.fun* 🟣\n`,
    `*Mint Address*\n\`${mintAddress}\`\n`,
    `*Transaction*\n\`${shortTx}\`\n`,
    `*Pump.fun Page*\n${viewUrl}\n`,
    `*Status:* Confirmed`,
    `*Time:* ${new Date(timestamp).toUTCString()}`,
  ];
  if (creatorBuySol > 0) {
    lines.push(`\n*Creator Buy:* \`${creatorBuySol} SOL\` purchased on launch.`);
  }
  if (targetMcapUsd > 0) {
    lines.push(
      `*Auto-Sell Target:* \`$${targetMcapUsd.toLocaleString()}\` market cap — the bot will sell automatically when hit.`
    );
  }
  lines.push(`\nTap *Token Panel* to manage your token.`);
  return lines.join("\n");
}

export function raydiumSuccessMessage(
  mintAddress: string,
  txSignature: string,
  solscanUrl: string,
  timestamp: string,
  targetMcapUsd: number
): string {
  const shortTx = `${txSignature.slice(0, 8)}...${txSignature.slice(-8)}`;
  const raydiumUrl = `https://raydium.io/liquidity/create-pool/?token=${mintAddress}`;
  const lines: string[] = [
    `*Token Deployed — Ready for Raydium* 🔵\n`,
    `*Mint Address*\n\`${mintAddress}\`\n`,
    `*Transaction*\n\`${shortTx}\`\n`,
    `*Solscan*\n${solscanUrl}\n`,
    `*Status:* Confirmed`,
    `*Time:* ${new Date(timestamp).toUTCString()}\n`,
    `*Create Raydium Pool:*\n${raydiumUrl}`,
  ];
  if (targetMcapUsd > 0) {
    lines.push(
      `\n*Auto-Sell Target:* \`$${targetMcapUsd.toLocaleString()}\` market cap — the bot monitors DEX Screener and notifies you when hit.`
    );
  }
  lines.push(`\nTap *Token Panel* to manage your token.`);
  return lines.join("\n");
}

export function autoSellExecutedMessage(
  mintAddress: string,
  txSignature: string,
  marketCapUsd: number,
  solReceived?: number
): string {
  const shortTx = `${txSignature.slice(0, 8)}...${txSignature.slice(-8)}`;
  return (
    `*Auto-Sell Triggered* 💰\n\n` +
    `*Mint:* \`${mintAddress}\`\n` +
    `*Market Cap Hit:* \`$${marketCapUsd.toLocaleString()}\`\n` +
    (solReceived !== undefined ? `*SOL Received:* \`${solReceived.toFixed(4)} SOL\`\n` : ``) +
    `*Transaction:* \`${shortTx}\`\n\n` +
    `*Status:* All creator tokens sold.`
  );
}

export function autoSellNotifyMessage(
  mintAddress: string,
  marketCapUsd: number
): string {
  return (
    `*Market Cap Target Reached!* 🎯\n\n` +
    `*Mint:* \`${mintAddress}\`\n` +
    `*Current Market Cap:* \`$${marketCapUsd.toLocaleString()}\`\n\n` +
    `Your target has been reached. Use Token Panel to sell your tokens now.`
  );
}

export function errorMessage(err: string): string {
  return `*Deployment Failed*\n\n${err}\n\nUse /reset to start over or /launch to retry.`;
}

export function insufficientFundsMessage(
  balance: number,
  required: number,
  breakdown?: { label: string; sol: number }[]
): string {
  const needed = (required - balance).toFixed(4);
  const lines = [
    `*Insufficient Funds*\n`,
    `*Current balance:* \`${balance.toFixed(4)} SOL\``,
    `*Required:* \`${required.toFixed(4)} SOL\``,
    `*Still needed:* \`${needed} SOL\``,
  ];
  if (breakdown && breakdown.length > 0) {
    lines.push(`\n*Breakdown:*`);
    for (const item of breakdown) {
      lines.push(`• ${item.label}: \`${item.sol.toFixed(4)} SOL\``);
    }
  }
  lines.push(`\nDeposit SOL to your wallet and use /wallet to verify your balance.`);
  return lines.join("\n");
}

File: artifacts/api-server/bot/messages.py
from config import DEPLOYMENT_FEE, PLATFORM_FEE_USD, DEX_UPDATE_USD, DEX_BOOST_USD


def escape(text: str) -> str:
    for ch in r"\*_[]()~`>#+-=|{}.!":
        text = text.replace(ch, f"\\{ch}")
    return text


def main_menu_message() -> str:
    return (
        "*TokenLaunchBot* — Solana SPL Token Deployer\n\n"
        "Deploy tokens on Solana in minutes\\.\n\n"
        "*Commands*\n"
        "/create — Start token creation\n"
        "/wallet — Deployment wallet info\n"
        "/withdraw — Withdraw SOL\n"
        "/panel — Token control panel\n"
        "/review — Review deployment details\n"
        "/launch — Deploy your token\n"
        "/reset — Start over\n"
        "/help — Show help"
    )


def wallet_message(address: str, balance: float, key_configured: bool) -> str:
    status = "✅ Configured" if key_configured else "❌ Not configured"
    enough = "✅ Sufficient" if balance >= DEPLOYMENT_FEE else f"⚠️ Need {DEPLOYMENT_FEE - balance:.4f} more SOL"
    return (
        f"*💼 Deployment Wallet*\n\n"
        f"*Address*\n`{address}`\n\n"
        f"*Balance*\n`{balance:.4f} SOL` — {enough}\n\n"
        f"*Minimum Required*\n`{DEPLOYMENT_FEE} SOL`\n\n"
        f"*Private Key*\n`{status}`\n\n"
        f"*Network*\nSolana Mainnet\n\n"
        f"_This is a fixed deployment wallet\\._"
    )


def help_message() -> str:
    return (
        f"*TokenLaunchBot Help*\n\n"
        f"*Supported Network:* Solana \\(SPL Tokens\\)\n\n"
        f"*Minimum Wallet Balance:* `{DEPLOYMENT_FEE} SOL`\n"
        f"_Covers token creation, creator buy, and all on\\-chain fees\\._\n\n"
        f"*How it works:*\n"
        f"1\\. /create — Enter token details\n"
        f"2\\. /review — Check everything looks right\n"
        f"3\\. /launch — Choose launchpad, set creator buy & market cap target\n"
        f"4\\. /panel — Manage your token after launch\n\n"
        f"*Required fields:* Name, Symbol\n"
        f"*Optional:* Description, Logo, Website, Telegram, Twitter\n\n"
        f"The bot uses a single dedicated deployment wallet\\.\n"
        f"Use /wallet to check your balance\\.\n\n"
        f"*Support*\n"
        f"Contact [@devBernard](https://t\\.me/devBernard) on Telegram\\."
    )


def review_message(token: dict, wallet_address: str) -> str:
    lines = ["*📋 Deployment Review*\n", "*Token Details*"]
    lines.append(f"Name: `{token['name']}`")
    lines.append(f"Symbol: `{token['symbol']}`")
    lines.append("Supply: `1,000,000,000` \\(fixed\\)")
    lines.append("Decimals: `9` \\(fixed\\)")

    if token.get("description"):
        lines.append(f"Description: _{escape(token['description'])}_")
    if token.get("logo_url"):
        lines.append(f"Logo: {token['logo_url']}")
    if token.get("website"):
        lines.append(f"Website: {token['website']}")
    if token.get("telegram"):
        lines.append(f"Telegram: {token['telegram']}")
    if token.get("twitter"):
        lines.append(f"Twitter: {token['twitter']}")

    lines.append("\n*Authority Settings*")
    lines.append(f"Revoke Mint: {'✅ Yes' if token.get('revoke_mint') else '❌ No'}")
    lines.append(f"Revoke Freeze: {'✅ Yes' if token.get('revoke_freeze') else '❌ No'}")

    lines.append(f"\n*Wallet:* `{wallet_address}`")
    lines.append(f"*Network:* Solana Mainnet")
    lines.append("\nReady to launch? Use /launch to continue\\.")
    return "\n".join(lines)


def launchpad_select_message(name: str, symbol: str) -> str:
    return (
        f"*Launch {escape(name)}* \\({escape(symbol)}\\)\n\n"
        f"*What the bot does after launch:*\n\n"
        f"🎯 *Auto Market Cap Push* — Set a target and the bot monitors price every 30 seconds\\. "
        f"When your target is hit, it auto\\-sells your creator tokens to lock in gains\\.\n\n"
        f"📣 *Community Shill Engine* — Your token gets pushed to active Solana trading communities "
        f"and listed on DEX Screener for early visibility and volume\\.\n\n"
        f"─────────────────\n"
        f"Choose your launchpad:"
    )


def creator_buy_message(launchpad: str, balance: float) -> str:
    lp_name = "Pump\\.fun" if launchpad == "pumpfun" else "Raydium"
    return (
        f"*Creator Buy*\n\n"
        f"Launchpad: *{lp_name}*\n"
        f"Wallet balance: `{balance:.4f} SOL`\n\n"
        f"How much SOL do you want to invest as the creator?\n"
        f"_This buys your own token on launch\\._"
    )


def target_mcap_message(creator_buy_sol: float) -> str:
    buy_str = f"`{creator_buy_sol} SOL`" if creator_buy_sol > 0 else "_none_"
    return (
        f"*🎯 Set Your Market Cap Target*\n\n"
        f"Creator buy: {buy_str}\n\n"
        f"At what market cap should the bot auto\\-sell all your creator tokens?\n\n"
        f"_The bot monitors price every 30 seconds\\. Once your target is hit, it sells everything instantly\\._"
    )


def dex_options_message(
    prices: dict,
    balance: float,
    creator_buy_sol: float,
    target_mcap: float,
    dex_update: bool,
    dex_boost: bool,
) -> str:
    mcap_str = f"${int(target_mcap):,}" if target_mcap > 0 else "No auto\\-sell"
    sol_usd = prices["sol_usd"]
    platform_sol = PLATFORM_FEE_USD / sol_usd
    upd_sol = prices["update_sol"] if dex_update else 0.0
    bst_sol = prices["boost_sol"] if dex_boost else 0.0
    total_sol = 0.05 + platform_sol + creator_buy_sol + upd_sol + bst_sol
    total_usd = total_sol * sol_usd

    lines = [
        "*DEX Screener Options* \\(optional\\)\n",
        f"Creator buy: `{creator_buy_sol} SOL`" if creator_buy_sol > 0 else "Creator buy: _none_",
        f"Auto\\-sell at: `{mcap_str}`",
        f"Wallet: `{balance:.4f} SOL`\n",
        f"*DEX Update* — `{prices['update_sol']:.4f} SOL` \\(${DEX_UPDATE_USD}\\)",
        "_Adds your token info to DEX Screener_\n",
        f"*DEX Boost* — `{prices['boost_sol']:.4f} SOL` \\(${DEX_BOOST_USD}\\)",
        "_Boosts your token to the trending section_\n",
        f"_SOL rate: ${sol_usd:.2f}/SOL_\n",
        "*── Total Cost ──*",
        f"Token creation: `0\\.0500 SOL`",
        f"Platform fee: `{platform_sol:.4f} SOL` \\(${PLATFORM_FEE_USD}\\)",
    ]
    if creator_buy_sol > 0:
        lines.append(f"Creator buy: `{creator_buy_sol:.4f} SOL`")
    if dex_update:
        lines.append(f"DEX Update: `{upd_sol:.4f} SOL` \\(${DEX_UPDATE_USD}\\)")
    if dex_boost:
        lines.append(f"DEX Boost: `{bst_sol:.4f} SOL` \\(${DEX_BOOST_USD}\\)")
    lines.append(f"*Total: `{total_sol:.4f} SOL` \\(~${total_usd:.0f}\\)*")
    return "\n".join(lines)


def insufficient_funds_message(balance: float, required: float) -> str:
    needed = required - balance
    return (
        f"*⚠️ Insufficient Funds*\n\n"
        f"*Current balance:* `{balance:.4f} SOL`\n"
        f"*Required:* `{required:.4f} SOL`\n"
        f"*Still needed:* `{needed:.4f} SOL`\n\n"
        f"Deposit SOL to your wallet, then use /wallet to verify and /launch to try again\\."
    )


def deploying_message(launchpad: str) -> str:
    lp = "Pump\\.fun" if launchpad == "pumpfun" else "Raydium"
    return (
        f"*🚀 Deploying on {lp}\\.\\.\\.*\n\n"
        f"This may take 30–60 seconds\\. Please wait\\."
    )


def pumpfun_success_message(
    mint_address: str,
    tx_signature: str,
    view_url: str,
    timestamp: str,
    creator_buy_sol: float,
    target_mcap_usd: float,
) -> str:
    short_tx = f"{tx_signature[:8]}\\.\\.\\."
    lines = [
        "*🟣 Token Live on Pump\\.fun* 🚀\n",
        f"*Mint Address*\n`{mint_address}`\n",
        f"*Transaction*\n`{short_tx}`\n",
        f"*Pump\\.fun Page*\n{view_url}\n",
        f"*Status:* ✅ Confirmed",
        f"*Time:* {timestamp}",
    ]
    if creator_buy_sol > 0:
        lines.append(f"\n*Creator Buy:* `{creator_buy_sol} SOL` invested")
    if target_mcap_usd > 0:
        lines.append(
            f"\n*Auto\\-Sell Target:* `${int(target_mcap_usd):,}` — "
            "the bot monitors every 30s and will notify you when hit\\."
        )
    lines.append("\nUse /panel to manage your token\\.")
    return "\n".join(lines)


def raydium_success_message(
    mint_address: str,
    tx_signature: str,
    timestamp: str,
    target_mcap_usd: float,
) -> str:
    short_tx = f"{tx_signature[:8]}\\.\\.\\."
    lines = [
        "*🔵 Token Created for Raydium* 🚀\n",
        f"*Mint Address*\n`{mint_address}`\n",
        f"*Transaction*\n`{short_tx}`\n",
        f"*Status:* ✅ Confirmed",
        f"*Time:* {timestamp}\n",
        "*Next Step:*\nGo to [raydium\\.io](https://raydium.io/liquidity/create\\-pool/) to create a liquidity pool\\.",
    ]
    if target_mcap_usd > 0:
        lines.append(
            f"\n*Auto\\-Sell Target:* `${int(target_mcap_usd):,}` — "
            "the bot monitors DEX Screener every 30s\\."
        )
    lines.append("\nUse /panel to manage your token\\.")
    return "\n".join(lines)


def panel_message(mint_address: str, symbol: str, ui_balance: float) -> str:
    return (
        f"*🎛 Token Control Panel*\n\n"
        f"*Token:* `{symbol}`\n"
        f"*Mint:* `{mint_address}`\n"
        f"*Balance:* `{ui_balance:,.2f} {symbol}`\n\n"
        f"Choose an action:"
    )


def burn_confirm_message(symbol: str, ui_amount: float, portion: str) -> str:
    label = "half" if portion == "half" else "ALL"
    return (
        f"*🔥 Confirm Burn*\n\n"
        f"Burn *{label}* of your `{symbol}` tokens\\.\n\n"
        f"Amount: `{ui_amount:,.4f} {symbol}`\n\n"
        f"⚠️ *This is irreversible\\.*"
    )


def revoke_confirm_message(authority_type: str, symbol: str) -> str:
    label = "Mint" if authority_type == "mint" else "Freeze"
    action = "mint new tokens" if authority_type == "mint" else "freeze token accounts"
    return (
        f"*🔒 Confirm Revoke {label} Authority*\n\n"
        f"Token: `{symbol}`\n\n"
        f"⚠️ *This is permanent and cannot be undone\\.*\n\n"
        f"No one will ever be able to {action} again\\."
    )


def withdraw_review_message(to_address: str, amount: float, balance: float) -> str:
    return (
        f"*📤 Withdrawal Review*\n\n"
        f"*To:* `{to_address}`\n"
        f"*Amount:* `{amount} SOL`\n"
        f"*Current Balance:* `{balance:.4f} SOL`\n"
        f"*Remaining After:* `{balance - amount:.4f} SOL`\n\n"
        f"Confirm to send\\."
    )


def withdraw_success_message(to_address: str, amount: float, tx_sig: str) -> str:
    short_tx = f"{tx_sig[:8]}\\.\\.\\."
    return (
        f"*✅ Withdrawal Sent*\n\n"
        f"*To:* `{to_address}`\n"
        f"*Amount:* `{amount} SOL`\n"
        f"*Transaction:* `{short_tx}`\n"
        f"*Status:* Confirmed"
    )


def error_message(reason: str) -> str:
    return f"*❌ Error*\n\n{reason}\n\n_Use /reset to start over\\._"


def deposit_notification_message(deposited: float, new_balance: float, wallet: str) -> str:
    status = (
        "✅ You have enough SOL\\. Use /launch when ready\\."
        if new_balance >= 2.0
        else f"⚠️ Minimum 2 SOL required\\. Still need `{2.0 - new_balance:.4f} SOL`\\."
    )
    return (
        f"*💰 Deposit Received*\n\n"
        f"*Amount:* `+{deposited:.4f} SOL`\n"
        f"*New Balance:* `{new_balance:.4f} SOL`\n"
        f"*Wallet:* `{wallet}`\n\n"
        f"{status}"
    )


def mcap_alert_message(mint_address: str, current_mcap: float, target_mcap: float, symbol: str) -> str:
    return (
        f"*🎯 Market Cap Target Hit\\!*\n\n"
        f"*Token:* `{symbol}`\n"
        f"*Mint:* `{mint_address}`\n"
        f"*Current Market Cap:* `${current_mcap:,.0f}`\n"
        f"*Your Target:* `${target_mcap:,.0f}`\n\n"
        f"Use /panel to sell your tokens now\\."
    )


def mcap_autosell_message(mint_address: str, tx_sig: str, mcap: float, symbol: str) -> str:
    short_tx = f"{tx_sig[:8]}\\.\\.\\."
    return (
        f"*✅ Auto\\-Sell Executed\\!*\n\n"
        f"*Token:* `{symbol}`\n"
        f"*Market Cap:* `${mcap:,.0f}`\n"
        f"*Transaction:* `{short_tx}`\n\n"
        f"All creator tokens sold successfully\\."
    )
